ERCS/erc-7771.md

```diff
@@ -0,0 +1,647 @@
+
+Having mixins be the intermediaries of storage, ensures that storage is always used in a secure way. A module defines its associated storage, and a mixin that specializes in managing this storage. Other modules that require reading or writing to this storage can safely do so by using the mixin as an intermediary.
+
+### Module Communication
+
+A crucial matter of a modular system is how modules communicate with each other. In a traditional system, `ModuleA` would need to call a registry `AddressResolver` to get the address of `ModuleB`, and then call the module.
+
+If the function in `ModuleB` has access restrictions, the latter module would then need to check with `AddressResolver` again, to verify if the calling module is an authorized module of the system. A simple communication between modules in this case results in 3 `CALL` operations, and is thus very inefficient and expensive.
+
+Initially, one may think that router communications could be solved by having `ModuleA` cast itself as `ModuleB` to call the latter’s function, which would effectively resolve the need for an `AddressResolver`, since the caller would be the system itself.
+
+```
+contract ModuleA {
+  function communicateWithModuleB() public {
+    // Don't do this!
+    uint value = ModuleB(address(this)).doSomething();
+  }
+}
+```
+
+However, this approach looses information of `msg.sender`, since `ModuleB` would receive `ModuleA`'s address as the sender, instead of the user.
+
+Also, this still uses 1 `CALL` operation, which can be avoided, as we’ll see below.
+
+This `msg.sender` issue could be resolved by using `DELEGATECALL` instead of `CALL`,
+
+```
+contract ModuleA {
+  function communicateWithModuleB() public {
+    (bool success, bytes memory value)
+       = address(this).delegatecall(ModuleB.doSomething.selector);
+    // Decode value here...
+  }
+}
+```
+
+But this would not only be cumbersome, but unnecessary.
+
+In fact, intermodular communication in a router proxy does not even need a single `CALL`. The final method we propose is using mixins.
+
+```
+contract ModuleA is MixinB {
+    function doSomething() public {
+        MixinB.doSomething();
+    }
+}
+```
+
+`MixinB` is a utility that knows how to operate on the storage namsepace `StorageB`, used by `ModuleB`. All that `ModuleA` needs to do to communicate with `ModuleB`, is extend its mixin representative. This approach is not only secure, but does not require even a single `CALL` or `DELEGATECALL`.
+
+A system designed with this architecture delegates much of its logic to mixins, making modules be mostly interfaces. Such an architecture is proven to scale in a healthy manner, easily overcoming contract size limit constraints.
+
+The main factor of such scalability is the clear separation of storage definitions, storage actuators (mixins), and user facing interfaces.
+
+### Full System Implementation
+
+A complex system using the architecture described above can be found in Synthetix v3 code in Github.
```

Comment:
```suggestion
```

This is an external link and must be removed. If you'd like to include a reference implementation under a compatible license, you may do so in the assets directory.